plugins/nominal-connection-checker/src/hierarchy_validation.js

```diff
@@ -0,0 +1,168 @@
+ */
+'use strict';
+
+/**
+ * Validates the given hierarchy definition. Does checks for duplicate types,
+ * circular dependencies, etc. Errors are logged to the console (not thrown).
+ * @param {!Object} hierarchyDef The definition of the type hierarchy.
+ */
+export function validateHierarchy(hierarchyDef) {
+  if (!hierarchyDef ||
+      typeof hierarchyDef != 'object' ||
+      Array.isArray(hierarchyDef)) {
+    console.error('The hierarchy definition should be an object.');
+    return;
+  }
+
+  checkConflictingTypes(hierarchyDef);
+  checkSupersDefined(hierarchyDef);
+  checkCircularDependencies(hierarchyDef);
+}
+
+/**
+ * Checks the hierarchyDef for conflicting types.
+ * @param {!Object} hierarchyDef The definition of the type hierarchy.
+ */
+function checkConflictingTypes(hierarchyDef) {
+  const conflictMsg = 'The type name \'%s\' conflicts with the type name(s) %s';
+
+  // Map of caseless names to cased names.
+  const definedTypes = new Map();
+  // Map of original names to arrays of conflicting names.
+  const conflictingTypes = new Map();
+
+  for (const type of Object.keys(hierarchyDef)) {
+    const caselessType = type.toLowerCase();
+    if (definedTypes.has(caselessType)) {
+      const originalType = definedTypes.get(caselessType);
+      if (!conflictingTypes.has(originalType)) {
+        conflictingTypes.set(originalType, [type]);
+      } else {
+        conflictingTypes.get(originalType).push(type);
+      }
+    } else {
+      definedTypes.set(caselessType, type);
+    }
+  }
+
+  for (const [type, conflicts] of conflictingTypes) {
+    console.error(conflictMsg, type, conflicts);
+  }
+}
+
+/**
+ * Checks the hierarchy def for any super types (eg 'fulfills': ['type]) which
+ * are not defined as top-level types.
+ * @param {!Object} hierarchyDef The definition of the type hierarchy.
+ */
+function checkSupersDefined(hierarchyDef) {
+  const errorMsg = 'The type %s says it fulfills the type %s, but that type' +
+      ' is not defined';
+
+  const types = new Set();
+  const keys = Object.keys(hierarchyDef);
+  for (const type of keys) {
+    types.add(type.toLowerCase());
+  }
+  for (const type of keys) {
+    const typeInfo = hierarchyDef[type];
+    if (!typeInfo.fulfills) {
+      continue;
+    }
+    for (const superType of typeInfo.fulfills) {
+      if (!types.has(superType.toLowerCase())) {
+        console.error(errorMsg, type, superType);
+      }
+    }
+  }
+}
+
+/**
+ * Checks for any circular dependencies between types. Eg:
+ * 'typeA': {
+ *   'fulfills': ['typeB']
+ * },
+ * 'typeB': {
+ *   'fulfills': ['typeA']
+ * }
+ * @param {!Object} hierarchyDef The definition of the type hierarchy.
+ */
+function checkCircularDependencies(hierarchyDef) {
+  /**
+   * Logs an informative error about a cyclic dependency.
+   * @param {!Array<string>} cycleArray An array of all the types that have been
+   *     visited since our entry node.
+   */
+  function logCycle(cycleArray) {
```

Comment:
Why is this defined inside `checkCircularDependencies` instead of separately? It would not be exported in either case.

Reply:
Oh yeah I got used to doing this from working in Scheme hehe. The idea is that because the `logCycle` function is only used by the `checkCircularDependencies` function you define it as in an inner function so that (a) you remember that and (b) you can collapse the outer function and hide everything at once.

It's really just a code org thing, so idc how it ends up =) Do you want me to move the inner funcs to top-level definitions?

Reply:
I have a slight preference for having it at the top level. Let mek now which you pick and when you're ready to merge.

Reply:
I moved `logCycle` out and renamed it to `logCircularDependency` to be more clear. I left `searchForCyclesRec` as an inner function because I forgot it was a closure.

Should be gtg :D